src/ApiCompat/Microsoft.DotNet.ApiCompatibility/Rules/RuleRunner.cs

```diff
@@ -50,14 +51,20 @@ public IReadOnlyList<IEnumerable<CompatDifference>> Run<T>(ElementMapper<T> mapp
                 else if (mapper is MemberMapper mm)
                 {
                     if (mm.ShouldDiffElement(rightIndex))
+                    {
+                        // ContainingType Left and Right cannot be null, as otherwise, the above condition would be false.
+                        Debug.Assert(mm.ContainingType.Left != null);
+                        Debug.Assert(mm.ContainingType.Right[rightIndex] != null);
+
                         _context.RunOnMemberSymbolActions(
                             mm.Left,
                             mm.Right[rightIndex],
-                            mm.ContainingType.Left,
-                            mm.ContainingType.Right[rightIndex],
+                            mm.ContainingType.Left!,
```

Comment:
The reason why you still have to add a `!` here is because even though you have a `Debug.Assert` above, you are not saving that into a local field, so Roslyn can't guarantee that from the time you run `Debug.Assert` to when you use it the value has not changed. If you want to get rid of this (and also do a minor perf optimization) you can save the left and right above into a local, Debug.Assert on the locals, and then pass that to `RunOnMemberSymbolActions`

Reply:
Apparently this only works for debug builds as on release, the call sites are removed.